File: BOJ/Algorithms/DFS_BFS/2468.cpp
#include <iostream>
#include <deque>
#include <cstring>
#include <algorithm>
#define MAX 101
using namespace std;

int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int N;
int Matrix[MAX][MAX];
bool visit[MAX][MAX];

void BFS(int X, int Y, int SafeArea){
    deque<pair<int, int>> Q;
    Q.push_back({X, Y});
    visit[X][Y] = true;
    while(!Q.empty()){
        int x = Q.front().first;
        int y = Q.front().second;
        Q.pop_front();
        for(int i = 0 ; i < 4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx >= 0 && nx < N && ny >= 0 && ny < N){
                if(Matrix[nx][ny] >= SafeArea && visit[nx][ny] == false){
                    visit[nx][ny] = true;
                    Q.push_back({nx, ny});
                } 
            }
        }
    }
}

int main(){
    int N;
    cin >> N;
    int Min = 9999;
    int Max = -9999;
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++){
            cin >> Matrix[i][j];
            if(Matrix[i][j] > Max) Max = Matrix[i][j];
            if(Matrix[i][j] < Min) Min = Matrix[i][j];
        }
    }
    int safeArea_Max = Min;
    for(int i = Min; i <= Max; i++){
        int tmp = 0;
        memset(visit, false, sizeof(visit));
        for(int j = 0; j < N; j++){
            for(int k = 0; k < N; k++){
                if(Matrix[j][k] >= i && !visit[j][k]){
                    BFS(j, k,i);
                    tmp += 1;
                }
            }
        }
        if(tmp > safeArea_Max) safeArea_Max = tmp;
    }
    cout << safeArea_Max << '\n';
}

File: BOJ/Algorithms/Graph/DFS_BFS/2468.py
import sys ; input = lambda : sys.stdin.readline().rstrip()
from collections import deque

dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def BFS(X, Y, safeArea):
    Q = deque()
    Q.append((X, Y))
    visit[X][Y] = True
    while(Q):
        x, y = Q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < N and 0 <= ny < N and visit[nx][ny] == False and matrix[nx][ny] >= safeArea:
                visit[nx][ny] = True
                Q.append((nx, ny))

N = int(input())
matrix = [list(map(int, input().split())) for i in range(N)]
minValue = min(map(min, matrix))
maxValue = max(map(max, matrix))

safeArea_Max = minValue

for safeArea in range(minValue, maxValue+1):
    tmp = 0
    visit = [[False]*N for i in range(N)]
    for i in range(N):
        for j in range(N):
            if matrix[i][j] >= safeArea and visit[i][j] == False:
                BFS(i, j, safeArea)
                tmp += 1
    if tmp > safeArea_Max:
        safeArea_Max = tmp
print(safeArea_Max)